hibernate-core/src/main/java/org/hibernate/internal/NonContextualJdbcConnectionAccess.java

```diff
@@ -51,10 +55,31 @@ public Connection obtainConnection() throws SQLException {
 			);
 			listener.jdbcConnectionAcquisitionEnd();
 		}
+
+		try {
+			session.afterObtainConnection( connection );
+		}
+		catch (SQLException e) {
+			try {
+				releaseConnection( connection );
+			}
+			catch (SQLException re) {
+				e.addSuppressed( re );
+			}
+			throw e;
+		}
+		return connection;
 	}
 
 	@Override
 	public void releaseConnection(Connection connection) throws SQLException {
+		try {
+			session.beforeReleaseConnection( connection );
+		}
+		catch (SQLException e) {
+			log.warn( "Error before releasing JDBC connection", e );
+		}
```

Comment:
@sebersole I'm anticipating that you might object to what I'm doing here. If so, I need a better suggestion.

Reply:
There used to be a way that users specified the type of tenancy, but I think that got dropped.  In a way, this PR adds that distinction back.

Another option would be to infer this when there is a ConnnectionProvider (and NOT a MultiTenantConnectionProvider) with a configured CurrentTenantIdentifierResolver.

Trying to get away from the need for a setting here.  Or is there some nuance I'm missing that makes you think we need it?

Reply:
> So the idea is that a user would not need MultiTenantConnectionProvider right? 

Yes.

> They could just do the normal `ConnectionProvider` set up and, in conjunction with `CurrentTenantIdentifierResolver`, we'd just alter the schema/catalog used by the Session - right?

Yes.

> It would be better if there were an actual flag saying that the user wants schema/catalog tenancy.

That's ... pretty much what I have implemented here, I believe. I just called the property something different.

> Have you looked at incorporating this into the `LogicalConnection` for the Session? Naively perhaps, that would seem like a better place.

I have not. Will take a look at that. 

OTOH, I think I would prefer this not be too deep down in the layers. It's a fairly high-level construct.

Reply:
> OTOH, I think I would prefer this not be too deep down in the layers. It's a fairly high-level construct.

I'd argue that this is far deeper than LogicalConnection, which is one of the reasons I mentioned it.  Plus LogicalConnection is the "eye of the needle" (in the good way) for Connection-related access for the Session.

Reply:
> There used to be a way that users specified the type of tenancy, but I think that got dropped. In a way, this PR adds that distinction back.

Yeah, I thought the same.

Reply:
> Trying to get away from the need for a setting here. Or is there some nuance I'm missing that makes you think we need it?

I suppose it would be in principle possible to default to schema-based multi-tenancy in the case that:

1. there's a defined tenant id,
2. there's no `@TenantId` annotation anywhere, and
3. there's no `MultiTenantConnectionProvider`.

So schema-based multi-tenancy would become some sort of default way we do multi-tenancy.

Dunno if I love that, however.

Reply:
> I'd argue that this is far deeper than LogicalConnection, which is one of the reasons I mentioned it.

OK, I'll look into it.

Reply:
> Wouldn't it be a reasonable use case to use both datasource-based and schema-based multitenancy?

Well this is already _easily_ possible by just implementing your own `MultiTenantConnectionProvider`.

For such complicated scenarios, I think it's totally reasonable that you have to write a bit of code.

The problem I'm trying to solve here is to have a way to do schema-based multitenancy with _any_ regular-ol' `ConnectionProvider`, without having to provide a `MultiTenantConnectionProvider`.

Reply:
> My point was that if you introduce hibernate.tenancy.style=schema, you'll need to take into account that there's a custom style or whatever you want to call it, that goes beyond just datasource/schema/discriminator

I think this is a good point, and goes to what I suggested with the 3 situations.  Phrased in the terms here we have:

1. column-based (`@TenantId`)
2. explicit `MultiTenantConnectionProvider`-based
3. `ConnectionProvider` + "adjustment" of the Connection for the Session

> I get you're trying to make simple things simpler, I'd be wary about making complex things more complex or more confusing, because that looks like a very real risk here.

I'm not following this...  What is the complex case that you think is getting more complex here?